File: programs/chronos/src/instructions/daemon_create.rs
use {
    crate::state::*,
    anchor_lang::{prelude::*, solana_program::system_program},
    std::mem::size_of,
};

#[derive(Accounts)]
#[instruction(bump: u8)]
pub struct DaemonCreate<'info> {
    #[account(
        init,
        seeds = [SEED_DAEMON, signer.key().as_ref()],
        bump = bump,
        payer = signer,
        space = 8 + size_of::<Daemon>(),
    )]
    pub daemon: Account<'info, Daemon>,

    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(ctx: Context<DaemonCreate>, bump: u8) -> ProgramResult {
    // Get accounts.
    let daemon = &mut ctx.accounts.daemon;
    let signer = &ctx.accounts.signer;

    // Initialize daemon account.
    daemon.owner = signer.key();
    daemon.bump = bump;

    return Ok(());
}

File: programs/chronos/client/src/instruction/daemonCreate.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { Chronos } from "../idl";
import { Account } from "../account";

export type DaemonCreateArgs = {
  owner: PublicKey;
};

export class DaemonCreate {
  private account: Account;
  private chronos: Program<Chronos>;

  constructor(account: Account, chronos: Program<Chronos>) {
    this.account = account;
    this.chronos = chronos;
  }

  public async daemonCreate({
    owner,
  }: DaemonCreateArgs): Promise<TransactionInstruction> {
    const daemonPDA = await this.account.daemon.pda(owner);
    return this.chronos.instruction.daemonCreate(daemonPDA.bump, {
      accounts: {
        daemon: daemonPDA.address,
        owner: owner,
        systemProgram: SystemProgram.programId,
      },
    });
  }
}
